compiler/passes/analyses/track_volume.py

```diff
@@ -0,0 +1,215 @@
+import networkx as nx
+
+from compiler.data_structures import Program
+from compiler.passes.analyses.bs_analysis import BSAnalysis
+from compiler.data_structures.ir import *
+
+
+class VolumeTracker(BSAnalysis):
+
+    # Global variables. They need to be delcared out here in order to be accessible via self.*
+    variable_volume = dict();                # The dict that will store the volumes of relevent variables
+    violation_found = False
+    _program = None;
+
+    def __init__(self):
+        super().__init__("Volume Tracking") # Sets the name to Volume Tracking?
+        
+
+    def analyze(self, program: Program) -> dict: # The main function of the class
+
+        self._program = program
+
+        for root in program.functions:
+
+            for node, data in program.bb_graph.nodes(data=True):
+
+                for i in program.functions[root]['blocks'][node].instructions:
+
+                    handle(self, i)
+
+                    if (self.violation_found):
+                        #print("Exiting...")
+                        break;
+
+            if (self.violation_found):
+                        #print("Exiting...")
+                        break;
+
+        return {'name': self.name, 'result': [self.violation_found, self.variable_volume]} # Returns the relevent results
+
+
+
+def handle(self, instruction : IR): # The meat of the logic. Just ferries out the functions based on the type of instruction.

```

Comment:
This and the following function should be declared within the class.

You'll need to call each method as `self.<method>`, and remove the `self` parameter from the call, as the calling class instance implicitly sends itself as the first parameter.